File: SmartStandControl/Alarm.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace arduinoControlTest
{
    enum Repeat
    {
        NEVER,
        DAILY,
        WEEKLY,
        WEEKDAY,
        WEEKEND,
        MONTHLY,
        YEARLY
    }

    struct AlarmItem
    {
        public string name;
        public string message;
        public DateTime time;

        public bool noise;
        public bool ligth;
        public bool onArduino;

        public Repeat repeat;

        public override string ToString()
        {
            string text = "CreateAlarm:";
            text += message + ":";
            text += time.Year + ":";
            text += time.Month + ":";
            text += time.Day + ":";
            text += time.Hour + ":";
            text += time.Minute + ":";
            text += noise + ":";
            text += ligth + ":";
            text += repeat;
            return text;
        }

        public AlarmItem(string name, DateTime time, bool noise, bool ligth, bool onArduino, Repeat repeat, string message = "")
        {
            this.name = name;
            this.message = message;
            this.time = time;

            this.noise = noise;
            this.ligth = ligth;
            this.onArduino = onArduino;
            this.repeat = repeat;
        }
    }

    internal class Alarm
    {
        private const string filePath = "";

        public List<AlarmItem> items;
        public List<ListViewItem> listViewItems;

        public Alarm() 
        {
            items = new List<AlarmItem>();
            listViewItems = new List<ListViewItem>();
        }

        ~Alarm()
        {
            Save();
        }

        public void Load(string path = filePath)
        {
            AddItem(new AlarmItem());
        }

        public void Save(string path = filePath)
        {

        }

        public void AddItem(AlarmItem item)
        {
            items.Add(item);
            listViewItems.Add(new ListViewItem(new string[] { item.name, item.time.ToString() }));
        }
    }
}

File: SmartStand/main/alarm.cpp
#include "alarm.h"
#include "Arduino.h"

 AlarmItem::AlarmItem(DateTime time, String message, bool sound, bool led, Repeat repeat)
 {
   this->time = time;
   this->message = message;
   this->sound = sound;
   this->led = led;
   this->repeat = repeat;

   this->done = false;
 }

void Alarm::addAlarmItem(AlarmItem* alarmItem, int idx)
{
  alarmItems[idx] = alarmItem;
}

AlarmItem* Alarm::checkCurrentAlarm(DateTime now) 
{
  for(int idx = 0; idx < 5; ++idx)
  {
    AlarmItem* item = alarmItems[idx];
    if(equal(now, item->time, item->repeat) == 0)
      return item;
  }

  return nullptr;
}

int Alarm::equal(DateTime time1, DateTime time2, Repeat repeat) //works only on equal
{
  if(time1.minute() != time2.minute())
    return -1;

  if(time1.hour() != time2.hour())
    return -1;

  if(time1.day() != time2.day())
  {
    if(repeat == DAILY)
      return 0;

    if(repeat == WEEKLY)
      if(time1.dayOfTheWeek() == time2.dayOfTheWeek())
        return 0;
    
    return -1;
  }

  if(time1.month() != time2.month())
  {
    if(repeat == MONTHLY)
      return 0;
    else
      return -1;
  }

  if(time1.year() != time2.year())
  {
    if(repeat == YEARLY)
      return 0;
    else
      return -1;
  }

  if(repeat == NEVER)
    return 0;

  return -1;
}